saexec/saexec_test.go

```diff
@@ -0,0 +1,741 @@
+	"github.com/stretchr/testify/require"
+	"go.uber.org/goleak"
+
+	"github.com/ava-labs/strevm/blocks"
+	"github.com/ava-labs/strevm/blocks/blockstest"
+	"github.com/ava-labs/strevm/cmputils"
+	"github.com/ava-labs/strevm/gastime"
+	"github.com/ava-labs/strevm/hook"
+	saehookstest "github.com/ava-labs/strevm/hook/hookstest"
+	"github.com/ava-labs/strevm/proxytime"
+	"github.com/ava-labs/strevm/saetest"
+	"github.com/ava-labs/strevm/saetest/escrow"
+)
+
+func TestMain(m *testing.M) {
+	goleak.VerifyTestMain(
+		m,
+		goleak.IgnoreCurrent(),
+		// Despite the call to [snapshot.Tree.Disable] in [Executor.Close], this
+		// still leaks at shutdown. This is acceptable as we only ever have one
+		// [Executor], which we expect to be running for the entire life of the
+		// process.
+		goleak.IgnoreTopFunction("github.com/ava-labs/libevm/core/state/snapshot.(*diskLayer).generate"),
+	)
+}
+
+// SUT is the system under test, primarily the [Executor].
+type SUT struct {
+	*Executor
+	chain  *blockstest.ChainBuilder
+	wallet *saetest.Wallet
+	logger logging.Logger
+}
+
+// newSUT returns a new SUT. Any >= [logging.Error] on the logger will also
+// cancel the returned context, which is useful when waiting for blocks that
+// can never finish execution because of an error.
+func newSUT(tb testing.TB, hooks hook.Points) (context.Context, SUT) {
+	tb.Helper()
+
+	logger := saetest.NewTBLogger(tb, logging.Warn)
+	ctx := logger.CancelOnError(tb.Context())
+
+	config := params.AllDevChainProtocolChanges
+	db := rawdb.NewMemoryDatabase()
+	tdbConfig := &triedb.Config{}
+
+	wallet := saetest.NewUNSAFEWallet(tb, 1, types.LatestSigner(config))
+	alloc := saetest.MaxAllocFor(wallet.Addresses()...)
+	genesis := blockstest.NewGenesis(tb, db, config, alloc, blockstest.WithTrieDBConfig(tdbConfig))
+
+	chain := blockstest.NewChainBuilder(genesis)
+	chain.SetDefaultOptions(blockstest.WithBlockOptions(
+		blockstest.WithLogger(logger)),
+	)
+	src := BlockSource(func(h common.Hash, n uint64) *blocks.Block {
+		b, ok := chain.GetBlock(h, n)
+		if !ok {
+			return nil
+		}
+		return b
+	})
+
+	e, err := New(genesis, src, config, db, tdbConfig, hooks, logger)
+	require.NoError(tb, err, "New()")
+	tb.Cleanup(e.Close)
+
+	return ctx, SUT{
+		Executor: e,
+		chain:    chain,
+		wallet:   wallet,
+		logger:   logger,
+	}
+}
+
+// timeNotThreadsafe returns a clone of the gas clock that times execution. It
+// is only safe to call when all blocks passed to [Executor.Enqueue]
+// have been executed.
+func (e *Executor) timeNotThreadsafe() *gastime.Time {
+	return e.gasClock.Clone()
+}
+
+func defaultHooks() *saehookstest.Stub {
+	return &saehookstest.Stub{Target: 1e6}
+}
+
+func TestImmediateShutdownNonBlocking(t *testing.T) {
+	newSUT(t, defaultHooks()) // calls [Executor.Close] in test cleanup
+}
+
+func TestExecutionSynchronisation(t *testing.T) {
+	ctx, sut := newSUT(t, defaultHooks())
+	e, chain := sut.Executor, sut.chain
+
+	for range uint64(10) {
```

Comment:
```suggestion
	for range 10 {
```

Reply:
Ahh left over from before I had the `saetest.Wallet` to handle nonces for me. Thanks for catching this!